### PR TITLE
CONTRACTS: represent loop nesting structure as a graph

### PR DESCRIPTION
**Pull request by remi-delmas-3000**
Utility class for loop contracts instrumentation.

Co-authored with @qinheping

Extracted from https://github.com/diffblue/cbmc/pull/7541, can only be tested once all features are in place.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: src/goto-instrument/contracts/dynamic-frames/dfcc_loop_nesting_graph.h [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+class messaget;
+
+/// A graph node that stores information about a natural loop.
+struct dfcc_loop_nesting_graph_nodet : public graph_nodet<empty_edget>
+{
+public:
+  dfcc_loop_nesting_graph_nodet(
+    const goto_programt::targett head,
+    const goto_programt::targett latch,
+    const loopt instructions);
```
Comment (tautschnig) [inline src/goto-instrument/contracts/dynamic-frames/dfcc_loop_nesting_graph.h]: Do you actually need to pass this by value?

File: src/goto-instrument/contracts/dynamic-frames/dfcc_loop_nesting_graph.h [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+  /// Loop head instruction
+  goto_programt::targett head;
+
+  /// Loop latch instruction
+  goto_programt::targett latch;
+
+  /// Set of loop instructions
+  loopt instructions;
```
Comment (tautschnig) [inline src/goto-instrument/contracts/dynamic-frames/dfcc_loop_nesting_graph.h:41]: As above: should this perhaps be a reference?

File: src/goto-instrument/contracts/dynamic-frames/dfcc_loop_nesting_graph.cpp [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+        t->location_number > loop_latch->location_number)
+        loop_latch = t;
+    }
+
+    if(loop_latch == loop_head)
+    {
+      log.error() << "Could not find latch instruction of loop starting at: "
+                  << loop_head->source_location() << messaget::eom;
+      throw analysis_exceptiont("Could not find loop latch instruction");
+    }
```
Comment (tautschnig) [inline src/goto-instrument/contracts/dynamic-frames/dfcc_loop_nesting_graph.cpp]: I'd expect a `DATA_INVARIANT` here, unless this is a scenario that's actually possible in GOTO programs.

**Review by tautschnig — APPROVED** [review]

File: src/goto-instrument/contracts/dynamic-frames/dfcc_loop_nesting_graph.h (hunk 1 above)
Comment (remi-delmas-3000) [reply src/goto-instrument/contracts/dynamic-frames/dfcc_loop_nesting_graph.h]: No! fixed.

File: src/goto-instrument/contracts/dynamic-frames/dfcc_loop_nesting_graph.h (hunk 2 above)
Comment (remi-delmas-3000) [reply src/goto-instrument/contracts/dynamic-frames/dfcc_loop_nesting_graph.h:41]: An instance of this class might outlive the `loopt` ref received by the constructor, so having the `loopt` stored by value here seemed like the safer option.

File: src/goto-instrument/contracts/dynamic-frames/dfcc_loop_nesting_graph.cpp (hunk 3 above)
Comment (remi-delmas-3000) [reply src/goto-instrument/contracts/dynamic-frames/dfcc_loop_nesting_graph.cpp]: Agreed and fixed.

**Comment by tautschnig** [comment]
This now just needs fixes to Doxygen.

**Pull request merged**